File: python_packages/habana_frameworks/torch/hpu/random.py
###############################################################################
#
#  Copyright (c) 2021-2025 Intel Corporation
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#
###############################################################################

from collections.abc import Iterable

import habana_frameworks.torch._core_C as htcore

import torch
from torch import Tensor

from ._utils import _get_device_index

# Keeping default_generators as a list with single instance to keep aligned to cuda
default_generators: list[torch._C.Generator] = [htcore._get_default_generator()]


def _default_generator():
    return default_generators[0]


__all__ = [
    "get_rng_state",
    "get_rng_state_all",
    "set_rng_state",
    "set_rng_state_all",
    "manual_seed",
    "manual_seed_all",
    "seed",
    "seed_all",
    "initial_seed",
]


def get_rng_state(device: int | str | torch.device = "hpu") -> Tensor:
    device_index = _get_device_index(device, optional=True)
    if device_index != 0:
        raise RuntimeError("hpu get_rng_state supports only device 0")
    return _default_generator().get_state()


def set_rng_state(new_state: torch.Tensor, device: int | str | torch.device = "hpu") -> None:
    device_index = _get_device_index(device, optional=True)
    if device_index != 0:
        raise RuntimeError("hpu set_rng_state supports only device 0")
    _default_generator().set_state(new_state)


def manual_seed(seed) -> torch._C.Generator:
    seed = int(seed)
    return _default_generator().manual_seed(seed)


def seed() -> int:
    return _default_generator().seed()


def initial_seed() -> int:
    return _default_generator().initial_seed()


def get_rng_state_all() -> list[Tensor]:
    return [get_rng_state(0)]


def set_rng_state_all(new_states: Iterable[Tensor]) -> None:
    if len(new_states) != 1:
        raise RuntimeError("hpu set_rng_state_all supports only states len of 1")
    for state in new_states:
        set_rng_state(state, 0)


def manual_seed_all(seed: int) -> None:
    manual_seed(seed)


def seed_all() -> None:
    _default_generator().seed()
    _default_generator().initial_seed()

File: backend/random.cpp
/**
 * Copyright (c) 2021-2024 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "random.h"
#include <ATen/CPUGeneratorImpl.h>
#include <ATen/core/Generator.h>
#include "habana_kernels/lazy_kernels.h"
#include "habana_lazy/lazy_executor.h"
namespace habana {
namespace detail {
// Getting the HPU worker generator instance
at::Generator& getDefaultHPUGenerator() {
  static auto default_gen_hpu = createHPUGenerator();
  return default_gen_hpu;
}

// Utility to create a CPUGeneratorImpl. Returns a shared_ptr
at::Generator createHPUGenerator() {
  auto default_cpu_gen = at::detail::getDefaultCPUGenerator();
  auto gen =
      at::make_generator<at::CPUGeneratorImpl>(default_cpu_gen.current_seed());
  return gen;
}
} // namespace detail

uint32_t get_seed_hpu(const std::optional<at::Generator>& gen) {
  at::CPUGeneratorImpl* generator =
      at::get_generator_or_default<at::CPUGeneratorImpl>(
          gen, detail::getDefaultHPUGenerator());

  auto context = habana_lazy::get_device_lazy_execution_context();
  if (context->getDryRun()) {
    return 0;
  }
  // Acquire lock when using random generators
  std::lock_guard<std::mutex> lock(generator->mutex_);
  return generator->random();
}

at::Tensor get_seed_tensor_hpu(const std::optional<at::Generator>& gen) {
  int seed = get_seed_hpu(gen);
  at::Tensor seed_tensor = at::tensor(seed);
  auto t = habana_lazy::append_to_batch_h2d_list(seed_tensor);
  auto context = habana_lazy::get_device_lazy_execution_context();
  if (context->getCapturing()) {
    habana_lazy::HbLazyTensor hb_tensor = habana_lazy::GetHbLazyTensor(t);
    hb_tensor.getDataPtr()->is_random_seed_tensor = true;
    context->getSeedTensorMap()[hb_tensor.getDataPtr()->unique_id] = gen;
  }
  return t;
}

} // namespace habana
